src/transform/src/semijoin_idempotence.rs

```diff
@@ -0,0 +1,485 @@
+        if distinct_on_keys_of(&inputs[0], &ltr)
+            && input_mapper.input_arity(0) == equivalences.len()
+        {
+            for mut candidate in semijoin_replacements(&inputs[0], bindings, rebindings) {
+                if ids1.contains(&candidate.id) {
+                    if let Some(permutation) = validate_replacement(&rtl, &mut candidate) {
+                        inputs[0] = candidate.replacement.project(permutation);
+                        *implementation = mz_expr::JoinImplementation::Unimplemented;
+
+                        // Take a moment to think about pushing down `IS NOT NULL` tests.
+                        let typ0 = inputs[0].typ().column_types;
+                        let typ1 = inputs[1].typ().column_types;
+                        let mut is_not_nulls = Vec::new();
+                        for (col1, col0) in rtl.iter() {
+                            if !typ0[*col0].nullable && typ1[*col1].nullable {
+                                use mz_expr::UnaryFunc;
+                                is_not_nulls.push(
+                                    MirScalarExpr::Column(*col1)
+                                        .call_unary(UnaryFunc::IsNull(mz_expr::func::IsNull))
+                                        .call_unary(UnaryFunc::Not(mz_expr::func::Not)),
+                                )
+                            }
+                        }
+                        if !is_not_nulls.is_empty() {
+                            inputs[1] = inputs[1].take_dangerous().filter(is_not_nulls);
+                        }
+
+                        // GTFO because things are now crazy.
+                        return;
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Evaluates the viability of a `candidate` to drive the replacement at `semijoin`.
+///
+/// Returns a projection to apply to `candidate.replacement` if everything checks out,
+/// as well as the equivalence between columns of semijoin and the replacement.
+fn validate_replacement(
+    map: &HashMap<usize, usize>,
+    candidate: &mut Replacement,
+) -> Option<Vec<usize>> {
+    if candidate.columns.len() == map.len()
+        && candidate
+            .columns
+            .iter()
+            .all(|(c0, c1, _c2)| map.get(c0) == Some(c1))
+    {
+        candidate.columns.sort_by_key(|(_, c, _)| *c);
+        Some(
+            candidate
+                .columns
+                .iter()
+                .map(|(_, _, c)| *c)
+                .collect::<Vec<_>>(),
+        )
+    } else {
+        None
+    }
+}
+
+/// A restricted form of a semijoin idempotence information.
+///
+/// This identifies two inputs to a semijoin, one of which is a `Get{id}` and the other an arbitrary relation.
+/// We also indicate which columns were equated, from each inputs, as well as a permutation to the columns of
+/// the latter which can allow us to substitute it in for the relation that offers this information.
```

Comment:
Should be cleared up. The module level documentation has also changed to introduce specific `A`, `B`, `C` roles that hint at who might be offering up this information (strictly speaking, it can be offered by expressions that are none of these, in pursuit of ultimately producing the information for `B`).

Reply:
Great, thank you very much! I'll continue studying the code tomorrow.